src/graph/validation/validate_neighbors.rs

```diff
@@ -0,0 +1,404 @@
+// License: MIT
+// Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+//! Methods for validating that all components in a [`ComponentGraph`] are
+//! connected correctly.
+
+use crate::{
+    component_category::CategoryPredicates, ComponentCategory, Edge, Error, InverterType, Node,
+};
+
+use super::ComponentGraphValidator;
+
+impl<N, E> ComponentGraphValidator<'_, N, E>
+where
+    N: Node,
+    E: Edge,
+{
+    pub(super) fn validate_root(&self) -> Result<(), Error> {
+        self.ensure_root(self.root)?;
+        self.ensure_not_leaf(self.root)?;
+        self.ensure_exclusive_successors(self.root)?;
+
+        Ok(())
+    }
+
+    pub(super) fn validate_meters(&self) -> Result<(), Error> {
+        for meter in self.cg.components().filter(|n| n.is_meter()) {
+            self.ensure_predecessor_categories(
+                meter,
+                &[ComponentCategory::Grid, ComponentCategory::Meter],
+            )?;
+            self.ensure_successor_not_categories(meter, &[ComponentCategory::Battery])?;
+        }
+        Ok(())
+    }
+
+    pub(super) fn validate_inverters(&self) -> Result<(), Error> {
+        for inverter in self.cg.components().filter(|n| n.is_inverter()) {
+            let ComponentCategory::Inverter(inverter_type) = inverter.category() else {
+                continue;
+            };
+
+            self.ensure_predecessor_categories(
+                inverter,
+                &[ComponentCategory::Meter, ComponentCategory::Grid],
+            )?;
+
+            match inverter_type {
+                InverterType::Battery => {
+                    self.ensure_not_leaf(inverter)?;
+                    self.ensure_successor_categories(inverter, &[ComponentCategory::Battery])?;
```

Comment:
Also in terms of resilience, is there anything preventing us to just emit an error and continue and assume the inverter is a "<whatever the successor is> inverter"? Most likely it was an error when building the component graph, it would be like duck-typing, if we use it as a "<whatever the successor is> inverter" and quacks as a "<whatever the successor is> inverter", why not continuing? If it doesn't quack as we want it to quack, we'll emit some other errors when using it and getting errors.

Reply:
My plan was to introduce a `Config` param and a new `ComponentGraph::try_new_with_config()` constructor, that would allow configuring various behaviours, like `validation.allow_disconnected_components`, `unspecified_inverters_are_battery_inverters`, `validation.warn_and_continue`, etc.  Then we can configure them per-site if required.